sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1160 @@
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        y : array, shape (n_samples,)
+            The corresponding training labels indices.
+
+        classes : array, shape (n_classes,), optional (default=None)
+            The non-singleton classes, encoded as integers in [0, n_classes).
+            If None (default), they will be inferred from ``y``.
+
+        Returns
+        -------
+        target_neighbors: array, shape (n_samples, n_neighbors)
+            An array of neighbors indices for each of the samples.
+        """
+
+        t_start = time.time()
+        if self.verbose:
+            print('[{}] Finding the target neighbors...'.format(
+                self.__class__.__name__))
+            sys.stdout.flush()
+
+        neighbors_params = self.neighbors_params
+        if neighbors_params is None:
+            neighbors_params = {}
+
+        neighbors_params.setdefault('n_jobs', self.n_jobs)
+        target_neighbors = _select_target_neighbors(
+            X, y, self.n_neighbors_, classes=classes, **neighbors_params)
+
+        if self.verbose:
+            print('[{}] Found the target neighbors in {:5.2f}s.'.format(
+                self.__class__.__name__, time.time() - t_start))
+
+        return target_neighbors
+
+    def _compute_grad_static(self, X, target_neighbors):
+        """Compute the gradient contributed by the target neighbors.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The training samples.
+
+        target_neighbors : array, shape (n_samples, n_neighbors)
+            The k nearest neighbors of each of the samples from the same class.
+
+        Returns
+        -------
+        grad_target_neighbors, shape (n_features, n_features)
+            An array with the sum of all outer products of
+            (sample, target_neighbor) pairs.
+        """
+
+        t_grad_static = time.time()
+        if self.verbose:
+            print('[{}] Computing static part of the gradient...'.format(
+                self.__class__.__name__))
+
+        n_samples, n_neighbors = target_neighbors.shape
+        row = np.repeat(range(n_samples), n_neighbors)
+        col = target_neighbors.ravel()
+        tn_graph = csr_matrix((np.ones(target_neighbors.size), (row, col)),
+                              shape=(n_samples, n_samples))
+        grad_target_neighbors = _sum_weighted_outer_differences(X, tn_graph)
+
+        if self.verbose:
+            t_grad_static = time.time() - t_grad_static
+            print('[{}] Computed static part of the gradient in {:5.2f}s.'
+                  .format(self.__class__.__name__, t_grad_static))
+
+        return grad_target_neighbors
+
+    def _callback(self, transformation):
+        """Called after each iteration of the optimizer.
+
+        Parameters
+        ----------
+        transformation : array, shape(n_components, n_features)
```

Comment:
check that it's indeed this, in NCA the callback is applied on raveled data, i.e.: 
```suggestion
        transformation : array, shape(n_components * n_features,)
```

Reply:
Done.